### DISCUSSION TITLE
ESP8266 SDCARD and UFILESYS

### DISCUSSION THREAD

**Discussion by benzino77** [arendst/Tasmota#11704] [Q&A] [answered]
Hi there!

I have a question regarding esp8266 board and support for sdcard and ufilesys.
*Case 1*
`user_config_override.h` includes:
```C
#define USE_SPI
// #define USE_UFILESYS
#define USE_SDCARD
```
then `pio run -e tasmota` compile new firmware without errors.

*Case 2*
`user_config_override.h` includes:
```C
//#define USE_SPI
#define USE_UFILESYS
//#define USE_SDCARD
```
then `pio run -e tasmota` compile new firmware without errors.

*Case 3*
`user_config_override.h` includes:
```C
#define USE_SPI
#define USE_UFILESYS
#define USE_SDCARD
```
then `pio run -e tasmota` fails with error:
```
/workspace/Tasmota/tasmota/tasmota.ino:81:19: fatal error: SDFAT.h: No such file or directory

***************************************************************
* Looking for SDFAT.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:SDFAT.h"
* Web  > https://platformio.org/lib/search?query=header:SDFAT.h
*
***************************************************************

 #include <SDFAT.h>
                   ^
compilation terminated.
*** [.pio/build/tasmota/src/tasmota.ino.cpp.o] Error 1
======================================= [FAILED] Took 10.71 seconds =======================================

Environment    Status    Duration
-------------  --------  ------------
tasmota        FAILED    00:00:10.710
================================== 1 failed, 0 succeeded in 00:00:10.710 ==================================
```

On ESP32 boards this combination is a default configuration (for odroid, m5, webcam) and compiles well. Is this configuration exclusive on esp8266 boards?

**Comment by sfromis** [comment]
It will only [try](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota.ino#L75-L92) to include `SDFAT.h` for the combination of `#define` for `USE_UFILESYS`, `ESP8266` and `USE_SDCARD`. Possibly, support needs to be in a particular build environment from [platformio_tasmota_env.ini](https://github.com/arendst/Tasmota/blob/development/platformio_tasmota_env.ini) for SDFAT being included in `lib_extra_dirs`

Generally, many things introduced with ESP32 will not be supported with ESP8266, but in this case the combination of `#ifdef` specifically deals with ESP8266.

**Reply by benzino77** [reply]
From `tasmota.ino` file perspective it looks like it is supported combination.

**Reply by sfromis** [reply]
Yeah, someone tried to do it, but the build setup seems to not have been amended to include the library. To save compile time, not all compiles include all libraries used by all selectable features. Therefore my reference to `lib_extra_dirs`, but I'm not at all well-versed in the details of amending the setup.

**Reply by benzino77** [reply]
tasmota/tasmota.ino:

```C
#ifdef USE_UFILESYS
  #ifdef ESP8266
    #include <LittleFS.h>
    #include <SPI.h>
    #ifdef USE_SDCARD
      #include <SD.h>
      #include <SDFAT.h>
    #endif  // USE_SDCARD
  #endif  // ESP8266
  #ifdef ESP32
    #include <LITTLEFS.h>
    #ifdef USE_SDCARD
      #include <SD.h>
    #endif  // USE_SDCARD
    #include "FFat.h"
    #include "FS.h"
  #endif  // ESP32
#endif  // USE_UFILESYS
```

Maybe I'm missing something but it is not possible to compile `SDCARD` support if `UFILESYS` is not included also. So my first *Case* works perfect because it does not reach the place where `SDFAT.h` is included...
In other words you can't have (in current code status) SDCARD support on ESP8266 devices.

**Reply by sfromis** [reply]
Well, SD card support does not make sense without it being used for something. AFAICT, `USE_SDCARD` is specifically for enabling SD card support for `USE_UFILESYS`. While there may be a compile issue, nothing in the structure would seem to exclude SD card support on ESP8266, and the code specifically has [different paths](https://github.com/arendst/Tasmota/issues/9739) for SD card support on ESP8266 and ESP32, based on different data structures.

**Reply by benzino77** [reply]
In current code state you can't have SDCARD on ESP8266 because of compile errors. I understand that it is not excluded by the code in any way, but because of compile errors you can't have it enabled -this is what I'm trying to say.

The background is that in TasmoCompiler I would like to have feature to select "SD Card/LittleFS" which will work the same for ESP8266 and ESP32 boards. In current code state I would have to distinct between esp8266 and esp32 and change description of the feature  and what is included in it (for ESP32 it will include littlefs and sdcard and for esp8266 only littlefs).

So the question is: Is SDCARD supported on esp8266 boards or not?

**Reply by sfromis** [reply]
My assumption is that if you fix the compile error (as you already found out how to do), then the existing code path to *specifically* support SD card on ESP8266 will do what's needed. You could also build on the case-agnostic Windows platform, but that's of course no solution.

**Reply by benzino77** [reply]
Well, the goal should be to fix it or make it compile without errors in Tasmota official source code not only my specific fork of it. I'm not experienced enough to decide whether changing from `SDFAT.h` to `SdFat.h` is the proper solution.
Should I rise an issue for that? Make a PR? Is this issue at all? 

Adrian @ascillato what is your opinion about that?

**Reply by sfromis** [reply]
Obviously, the goal is to not have compile errors. That's why I *already* suggested you to open an issue/PR to fix the file name casing issue.

**Reply by benzino77** [reply]
I haven't seen that post below ;)

**Comment by benzino77** [comment, marked as answer]
Changing line:
```C
#include <SDFAT.h>
```
to 
```C
#include <SdFat.h>
```

allows to compile without errors. I'm not sure whether this is a proper solution and this will work as expected.

Error description itself was the hint which leads me to this observation:
```
* CLI  > platformio lib search "header:SDFAT.h"
* Web  > https://platformio.org/lib/search?query=header:SDFAT.h
```

platformio search engine does not find `SDFAT.h` header, but it can find `SdFat.h` included in library maintained by Bill Greiman

**Reply by sfromis** [reply]
Since ESP8266Audio [has](https://github.com/arendst/Tasmota/blob/development/lib/lib_audio/ESP8266Audio/src/AudioFileSourceSD.h) `#include <SdFat.h>` (without using `USE_SDCARD`) that change to mixed case does seem to be quite plausibly correct. The issue may have popped up due to the illogical use of camel case for abbreviations, and case differences not being detected when building in the case-agnostic Windows environment. The number of people having tried to enable the ESP8266 path for SD card support may be quite low.

**Reply by sfromis** [reply]
I'd suggest opening an issue/PR about the mixup with casing of that file name.

**Comment by barbudor** [comment]
We've seen case differences between esp8266 and ESP32 like LittleFS vs LITTLEFS
So be sure to get the right one

**Reply by benzino77** [reply]
I've already opened an issue (based on @sformis suggestion) for that: #11724. I'm "too short" to decide how it should look like for a particular board line :)

**Reply by sfromis** [reply]
While I'm not involved in Tasmota firmware development, what you found looks just right to me.
